File: quickshop/app/src/main/java/com/avas/quickshop/MainViewModel.kt
package com.avas.quickshop

/// Created by Aabhash Shakya on 19/11/2025
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.avas.quickshop.features.login.data.AuthRepository
import com.avas.quickshop.flutter_utils.FlutterEvent
import com.avas.quickshop.flutter_utils.FlutterEventsRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class MainViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val flutterEventsRepository: FlutterEventsRepository
) : ViewModel() {


    private val _isUserLoggedIn = MutableStateFlow<Boolean?>(null)
    val isUserLoggedIn = _isUserLoggedIn.asStateFlow()
    val uuid = authRepository.uuidFlow

    private val _flutterEvent = MutableStateFlow<FlutterEvent?>(null)
    val flutterEvent: StateFlow<FlutterEvent?> = _flutterEvent

    init {
        observeFlutterEvents()
        viewModelScope.launch {
            getStartDestination()
        }
    }

    suspend fun getStartDestination() {
        _isUserLoggedIn.value = authRepository.isUserLoggedIn()
    }


    private fun observeFlutterEvents() {
        viewModelScope.launch {
            flutterEventsRepository.events.collect { event ->

                when (event) {

                    is FlutterEvent.CacheUUID -> {
                        Log.i("FlutterEvent", "Cached uuid")
                        authRepository.saveUUID(event.uuid)
                    }

                    is FlutterEvent.ProductSelected -> {
                        // You can handle this if needed
                        Log.i("FlutterEvent", "Payment complete")
                        _flutterEvent.value = event
                    }

                    is FlutterEvent.UserLoggedOut -> {
                        Log.i("FlutterEvent", "User logged out")
                        authRepository.logout()

                    }

                }
            }
        }
    }

    fun clearFlutterEvent() {
        _flutterEvent.value = null
    }
}

File: QuickShop-Ios/QuickShop-Ios/vm/MainViewModel.swift
//
//  MainViewModel.swift
//  quickshop-ios
//
//  Created by Aabhash Shakya on 20/11/2025.
//

import Foundation
import Combine

class MainViewModel: ObservableObject {
    @Published var uuid: String? = SessionManager.shared.uuid
    @Published var flutterEvent: FlutterEvent?
    
    func setEvent(_ event: FlutterEvent) {
        flutterEvent = event
    }
    
    private var cancellable: AnyCancellable?

        init() {
            // subscribe to repository
            cancellable = FlutterEventsRepository.shared.$latestEvent
                .receive(on: DispatchQueue.main)
                .sink { [weak self] event in
                    if let event = event {
                              self?.handleFlutterEvent(event)
                          }                }
        }
    
    func handleFlutterEvent(_ event: FlutterEvent) {
        switch event {
        case .productSelected(_):
            self.flutterEvent = event
        case .cacheUUID(let uuid):
            cacheUUID(uuid)
        case .userLoggedOut:
            print("User logged out")
            logout()
        }
    }
    
    func cacheUUID(_ newUUID: String) {
         SessionManager.shared.uuid = newUUID
         uuid = newUUID
     }

        func clearFlutterEvent() {
            flutterEvent = nil
        }
    
    func logout(){
        SessionManager.shared.uuid = nil
        uuid = nil
        
    }

}
